keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,982 @@
+making use of v3 features does not have a good experience with inaccurate information, or fields removed altogther.
+
+This transformation causes bugs, for example, when attempting to `explain` a field
+that is `nullable`, kubectl instead shows nothing, due to the lossy conversion
+wiping nullable fields.
+
+### Goals
+
+1. Provide the new richer type information specified by OpenAPI v3 within kubectl explain
+2. Have a more maintainable `text/template` based approach to printing
+3. Fallback to old `explain` implementation if cluster does not expose OpenAPI v3 data.
+4. Provide multiple new output formats for kubectl explain:
+    * human-readable plaintext
+    * markdown
+    * maybe others
+5. (Optional?) Allow users to specify their own templates for use with kubectl
+  explain (there may be interesting use cases for this)
+6. Improve discoverability of API Resources and endpoints, and provide a platform
+for richer information to be included in the future.
+
+### Non-Goals
+
+1. "Fix" openapi v3 to openapi v2 conversion
+  This is a non-goal for two reasons:
+    * These formats are not compatible, and there WILL be data loss and inaccuracy
+    * This negates the benefits of using OpenAPI v3 for the richer type information
+2. Provide general-purpose OpenAPI visualization.
+
+
+## Proposal
+
+
+### Basic Usage
+
+The following user experience should be possible with `kubectl explain`
+
+```shell
+kubectl explain pods.spec
+```
+
+Output should be familiar to users of today's `kubectl explain`, except new
+information from the OpenAPI v3 spec is now populated.
+
+Note: Feature during development will be gated by an experimental flag. The commands
+shown here elide the experimental flag for clarity.
+
+### Built-in Template Options
+#### Plaintext
+
+
+```shell
+kubectl explain pods
+```
+or
+
+```shell
+kubectl explain pods --output plaintext
+```
+
+The plaintext output format is the default and should be crafted to be as close
+as the existing `explain` output in use before this KEP.
+
+#### OpenAPIV3 (raw json)
+
+```shell
+kubectl explain pods --output openapiv3
+```
+
+To get raw OpenAPI v3 data for a certain resource today involves:
+1.) setting up kubectl proxy
+2.) fetching the correct path at `/openapi/v3/<group>/<version>`
+3.) filtering out unwanted results
+
+This command is useful not only for its convenience, but also other visualizations
+may be built upon the raw output if we opt not to support a first-class custom
+template solution in the future.
+
+#### HTML
+
+```shell
+kubectl explain pods --output html
+```
+
+Similarly to [godoc](https://pkg.go.dev), we suggest to provide a searchable,
+navigable, generated webpage for the kubernetes types of whatever cluster kubectl
+is talking to.
+
+Only the fields selected in the command line (and their subfields' types, etc) 
+will be included in the resultant page.
+
+If user types `kubectl explain --output html` with no specific target, then all types
```

Comment:
I think the html and md outputs are provisional for now, so I can remove this section, or mark it more clearly as simply an idea

Reply:
I'm in favour of us marking as a possible future idea rather than removing. Potentially more on level with the custom templates idea, which is also provisional? And adding another provisional idea about allowing resourceless invocation, which could apply to any template format in theory.

Reply:
marked the idea as provisional